### ISSUE TITLE
Extension stalls if polling when network is down

### ISSUE THREAD

**Issue by theychx** [alexduf/gnome-github-notifications#11]
Hi. Thank you for making this extension, it's really useful.

Sometimes during startup of my box, gnome-shell is up and running before the network is up. In this scenario the extension fails (without showing the exclamation-mark) and then afterwards does not do any further polling.
I'm using gnome-shell 3.26.

**Comment by alexduf** [comment]
Hi, thanks for the feedback!
Yes I have observed a similar behaviour when the network is lost mid-session. I haven't identified why as in theory all exceptions should be caught.

I'll leave this ticket open and if I find a solution I'll reply here.

Thanks @theychx

**Comment by alexduf** [comment]
Actually I'm wondering if my retry logic isn't the problem.

So I created the branch `remove-retry-logic` that I'm currently running. I'll check if that solves the issue and I'll push the fix if that appears to work. If you know how to checkout a branch and manually install an extension that would be great if you could test it as well!

Don't hesitate to ping me here if I don't reply, I may very well have forgotten about it :)

**Comment by theychx** [comment]
You didn't push anything to ```remove-retry-logic```. :smiley: 
I occurs to me that this might be the offending line: https://github.com/alexduf/gnome-github-notifications/blob/master/extension.js#L199
This hack seems to work (I have zero javascript or gnome api skills):
```
if (!response.response_body.data) {
  if (response.status_code > 400) {
    error('HTTP error:' + response.status_code);
  } else {
    error('Unknown network error');
  }
  this.planFetch(this.interval(), true);
  this.label.set_text('!');
  return;
```

**Comment by alexduf** [comment]
That's a good shout as well. I'll keep trying that branch, if that doesn't give me any result, I'll try forcing the planFetch in all the cases.

> I have zero javascript or gnome api skills

Neither do I 🙃

**Comment by mehdisadeghi** [comment]
@alexduf I think I might have the same problem and since you mentioned network errors, it could be that after standby the extension does not work anymore, unless I manually disable and re-enable it. I applied the hack from @theychx to see whether it works. I'll update later.

**Comment by Shelvak** [comment]
Hi guys, I changed the code because after a few hours Gh api returns `status_code: 6` I have no Idea what that means... At the moment I'm running this code:
```javascript
if (!response.response_body.data) {
          error('HTTP error:' + response.status_code);
          error('response error: ' + JSON.stringify(response));
          this.planFetch(this.interval(), true);
          this.label.set_text('!');
          return;                                                                                                                                                             
}
```
And in same cases I see that:
```
jun 07 20:00:33 rotsen-pc gnome-shell[595]: [GITHUB NOTIFICATIONS EXTENSION][ERROR] HTTP error:6                                                                             
jun 07 20:00:33 rotsen-pc gnome-shell[595]: [GITHUB NOTIFICATIONS EXTENSION][ERROR] response error: {} 
```
I'm testing the code and grepping  logs (for about 48hs now), It works really good =)

**Comment by alexduf** [comment]
That's really interesting information @Shelvak thanks!

I searched a bot and it looks like HTTP status bellow 100 are local errors:
https://github.com/GNOME/libsoup/blob/master/libsoup/soup-status.h#L26-L36

I guess I never thought about that, I'll patch and test today, if that works I'll release ASAP.

**Comment by alexduf** [comment]
I'll be running off this branch, if anyone feels adventurous, please do have a go at it:
https://github.com/alexduf/gnome-github-notifications/pull/14

**Comment by alexduf** [comment]
We could retry forever in that case, but I'm worried it may fall into that retry loop in cases that aren't necessary.

Maybe it's a good case to have an exponential retry policy

**Comment by Shelvak** [comment]
That's a good point. I'll code something at the weekend, and post here   :)

**Comment by alexduf** [comment]
Fixed by #14

**Comment by Shelvak** [comment]
Awesome =)  I didn't have time to code the retry function, when I have it I'll ping you =D Thanks for the awesome extension :+1:
